clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3562,122 +3562,126 @@ static void emitCPPTypeString(raw_ostream &OS, QualType Ty) {
-    ArgOS << TD->getQualifiedNameAsString();
-    break;
-  }
-  default:
-    Arg.print(P, ArgOS);
+  void VisitQualifiers(Qualifiers Quals) {
+    Quals.print(O, P, /*appendSpaceIfNotEmpty*/ true);
   }
-}
 
-static void printArguments(ASTContext &Ctx, raw_ostream &ArgOS,
-                           ArrayRef<TemplateArgument> Args,
-                           const PrintingPolicy &P) {
-  for (unsigned I = 0; I < Args.size(); I++) {
-    const TemplateArgument &Arg = Args[I];
+public:
+  SYCLKernelNameTypePrinter(raw_ostream &Out, PrintingPolicy &P)
+      : Out(Out), P(P), O(Buf) {}
 
-    // If argument is an empty pack argument, skip printing comma and argument.
-    if (Arg.getKind() == TemplateArgument::ArgKind::Pack && !Arg.pack_size())
-      continue;
+  ~SYCLKernelNameTypePrinter() { emitWithoutAnonNamespaces(Out, O.str()); }
+
+  void Visit(QualType T) {
+    if (T.isNull())
+      return;
 
-    if (I != 0)
-      ArgOS << ", ";
+    QualType CT = T.getCanonicalType();
+    VisitQualifiers(CT.getQualifiers());
 
-    printArgument(Ctx, ArgOS, Arg, P);
+    InnerTypeVisitor::Visit(CT.getTypePtr());
   }
-}
 
-static void printTemplateArguments(ASTContext &Ctx, raw_ostream &ArgOS,
-                                   ArrayRef<TemplateArgument> Args,
-                                   const PrintingPolicy &P) {
-  ArgOS << "<";
-  printArguments(Ctx, ArgOS, Args, P);
-  ArgOS << ">";
-}
+  void VisitType(const Type *T) {
+    O << QualType::getAsString(T, Qualifiers(), P);
+  }
 
-static void emitRecordType(raw_ostream &OS, QualType T, const CXXRecordDecl *RD,
-                           const PrintingPolicy &TypePolicy) {
-  SmallString<64> Buf;
-  llvm::raw_svector_ostream RecOS(Buf);
-  T.getCanonicalType().getQualifiers().print(RecOS, TypePolicy,
-                                             /*appendSpaceIfNotEmpty*/ true);
-  if (const auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
+  void Visit(const TemplateArgument &TA) {
+    if (TA.isNull())
+      return;
+    InnerTAVisitor::Visit(TA);
+  }
+
+  void VisitTagType(const TagType *T) {
+    TagDecl *RD = T->getDecl();
+    if (const auto *TSD = dyn_cast<ClassTemplateSpecializationDecl>(RD)) {
 
-    // Print template class name
-    TSD->printQualifiedName(RecOS, TypePolicy, /*WithGlobalNsPrefix*/ true);
+      // Print template class name
+      TSD->printQualifiedName(O, P, /*WithGlobalNsPrefix*/ true);
 
-    // Print template arguments substituting enumerators
-    ASTContext &Ctx = RD->getASTContext();
-    const TemplateArgumentList &Args = TSD->getTemplateArgs();
-    printTemplateArguments(Ctx, RecOS, Args.asArray(), TypePolicy);
+      ArrayRef<TemplateArgument> Args = TSD->getTemplateArgs().asArray();
+      O << "<";
+      printTemplateArgs(Args);
+      O << ">";
 
-    emitWithoutAnonNamespaces(OS, RecOS.str());
-    return;
+      return;
+    }
+    // Q: Next part of code results in printing of "class" keyword before class
```

Comment:
Can you give some examples on this?  It seems that the 'printing policy' that you pass (P?) could do SuppressTagKeyword as true (though, it SHOULD be true for C++ mode?) which should do that, right?

EDIT: I see you discovered this below :)  I don't see reason to add the 'class' keyword if it isn't necessary, it is likely in the tests as a byproduct of the previous printer and noone investigating why it happpened.